Add zigpy-ota OTA provider

TheJulianJES opened the pull request:
## Proposed change

This adds an OTA provider for the zigpy-ota repo.

### Additional information

The provider is enabled with the `stable` channel by default, but you can change the channel to `beta` or `dev`. The difference is explained in the README of the [zigpy-ota repo](https://github.com/zigpy/zigpy-ota).
```yaml
zha:
  # custom_quirks_path: /config/zha_custom_quirks
  zigpy_config:
    ota:
      # Default: stable channel is enabled automatically (no config needed)

      # Use beta channel instead of stable
      extra_providers:
        - type: zigpy_ota
          channel: beta  # or 'dev'
          override_previous: true

      # Use custom version file URL (for testing/self-hosted)
      extra_providers:
        - type: zigpy_ota
          url: https://example.org/custom/version.json
          override_previous: true
```

## TODO:

- ✅ ~~Make [zigpy-ota repo](https://github.com/zigpy/zigpy-ota) available~~
- ✅ ~~Actually test this in HA with real devices and OTAs~~
- ✅ ~~Consider trusting self-signed Hue and IKEA certificates for this provider?~~
  - ~~It's possible we wanna add remotely hosted images to the index that use these self-signed certificates
    Would look like this: https://github.com/TheJulianJES/zigpy/compare/tjj/add_zigpy-ota_provider_squashed...TheJulianJES:zigpy:tjj/add_zigpy-ota_provider_squashed_ssl~~
  - ~~If we were to add new entries (without changing the schema from `zigpy_v1` to `zigpy_v2`, for example), old zigpy versions would fail to download them. This might not be great, but it looks like we fail at debug level, so that's good.~~

### For possible future PRs, we may also want to consider these points:

- There's no throttling for image downloading if it fails. But at least it's there for fetching the index. Should this generally be added to zigpy?
  - It looks like we attempt to download compatible files every time a device sends an image request(?)
- For "trusted providers", it might be nice to avoid downloading a seemingly compatible image to verify, and only download the image if the user actually tries to install the update.

TheJulianJES committed:
Add zigpy-ota OTA provider

TheJulianJES committed:
Shorten docstring

TheJulianJES committed:
Test `__hash__` for zigpy-ota provider

Required for `combine_concurrent_calls`

TheJulianJES committed:
Allow Hue + IKEA self-signed certs for zigpy-ota

TheJulianJES committed:
Put certs into constants

TheJulianJES commented:
For also allowing Hue + IKEA self-signed certs when downloading images from the zigpy-ota provider, the changes would look like this: https://github.com/TheJulianJES/zigpy/compare/tjj/add_zigpy-ota_provider_squashed...TheJulianJES:zigpy:tjj/add_zigpy-ota_provider_squashed_ssl

I'm leaning towards just adding them, for now. If we don't end up using the remote/third-party image hosting feature at all, we can remove them later. But we'd be on the safe side for existing zigpy versions if we do end up using remote Hue/IKEA images. Or is this a bad idea to implement at all?

(We could also include custom/self-signed certs for downloading images in the index itself, since we only have a `firmware` key there so far. But I feel like that's over-doing it...)

EDIT: Added, but commits show up above.

puddly commented:
Yeah, let's just add them. There have only ever been two of these self-signed certificates. It's not like IKEA is going to man-in-the-middle Hue OTA images or vice versa.

MattWestb commented:
IKEAs OTA files is signed but not encrypted so if having one man in the middle and the system is downloading the firmware and sending it to one device with right man and image ID the device is downloading it OK but then start the flashing process its throwing the file then the signing is broken so shall being safe as long  not manipulating the OTA file metadata.
I dont knowing if HUE is signed / encrypted but i think so.
We have seen Aqara and tuya OTA files being wrong flashed on devices and broken them  but they cant have the OTA security implemented at all or not working OK.

TheJulianJES committed:
Also add `markdown_v1` schema to test files

This isn't used at all, but it makes the tests slightly more accurate if it's present in the mocked response

TheJulianJES committed:
Move `release_url` in test files

The order in the generated zigpy-ota was changed slightly, so we update the test files for the mocked response

TheJulianJES commented:
Tested this on a temporary test repo with the same file/release structure and it works as expected:
<img width="563" height="431" alt="Bildschirmfoto 2025-12-28 um 07 13 35" src="https://github.com/user-attachments/assets/81d224c9-7359-4d6c-a250-92e8fb0fb4e9" />
(and no, Hue images will not be part of the initial version for the zigpy-ota repo)

TheJulianJES commented:
From my side, this PR should be ready. I'll do some last checks later today and hopefully make the zigpy-ota repo public, so we can merge this. Since the HA beta is already on Monday, this would be very nice to get into HA before then.

TheJulianJES commented on zigpy/ota/providers.py:
```diff
@@ -660,6 +668,89 @@ async def _load_index(
+        async with session.get(self.url) as rsp:
+            version_data = await rsp.json(content_type=None)
+
+        # Extract the index URL from the version file
+        # Format: {"schemas": {"zigpy_v1": {"version": "...", "url": "..."}}}
+        index_url = version_data["schemas"]["zigpy_v1"]["url"]
```
In the future, it's possible that we provide a `zigpy_v2` schema and eventually drop hosting the `zigpy_v1` compatible index (though the zigpy-ota setup allows for us to host both versions of the index and just keep the old `zigpy_v1` version around, so this failure is somewhat unlikely).

If you were to run a very old version of zigpy/HA, we'd (currently) fail like this (at **debug level** only!):
```python
2025-12-28 07:34:58.737 DEBUG (MainThread) [zigpy.ota] Failed to load provider ZigpyOtaProvider(url='https://working_url_pointing_to_version_file_here.com/stable.json', channel=None, manufacturer_ids=())
Traceback (most recent call last):
  File "/Users/julian/PythonProjects/HomeAssistantcore/venv/lib/python3.13/site-packages/zigpy/ota/__init__.py", line 397, in get_ota_images
    index = await self._load_provider_index(provider)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/julian/PythonProjects/HomeAssistantcore/venv/lib/python3.13/site-packages/zigpy/util.py", line 467, in replacement
    return await tasks[key]
           ^^^^^^^^^^^^^^^^
  File "/Users/julian/PythonProjects/HomeAssistantcore/venv/lib/python3.13/site-packages/zigpy/ota/__init__.py", line 372, in _load_provider_index
    return await provider.load_index()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/julian/PythonProjects/HomeAssistantcore/venv/lib/python3.13/site-packages/zigpy/ota/providers.py", line 232, in load_index
    return [meta async for meta in self._load_index(session)]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/julian/PythonProjects/HomeAssistantcore/venv/lib/python3.13/site-packages/zigpy/ota/providers.py", line 726, in _load_index
    index_url = version_data["schemas"]["zigpy_v1"]["url"]
                ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: 'zigpy_v1'
```

We could also make it more explicit:
```python
# Extract the index URL from the version file
# Format: {"schemas": {"zigpy_v1": {"version": "...", "url": "..."}}}
schemas = version_data.get("schemas", {})

if "zigpy_v1" not in schemas:
    raise ValueError(
        f"Unsupported zigpy-ota index: 'zigpy_v1' schema not found. "
        f"Available schemas: {list(schemas.keys())}. "
        f"Please update zigpy to a newer version."
    )

index_url = schemas["zigpy_v1"]["url"]
```

Is this something we want, or is it fine to leave as is? In both cases, the error would only be visible at debug level when the provider is first loaded. It's not attempted to be loaded again until after 24 hours.

TheJulianJES replied:
I think the current behavior is fine. It's only ever a debug error **if** it ever happens anyway.
We can adjust this later if needed.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
+
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
+
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on tests/ota/test_ota_providers.py:
```diff
@@ -700,3 +700,127 @@ async def test_ota_fetch_size_and_checksum_validation(
+
+    for obj, meta in zip(index_obj["firmwares"], index, strict=True):
+        assert isinstance(meta, providers.RemoteOtaImageMetadata)
+        assert meta.url == obj.pop("binary_url")
+        assert meta.file_version == obj.pop("file_version")
+        assert meta.file_size == obj.pop("file_size")
+        assert meta.image_type == obj.pop("image_type")
+        assert meta.manufacturer_id == obj.pop("manufacturer_id")
+        assert meta.checksum == obj.pop("checksum")
+        assert meta.manufacturer_names == tuple(obj.pop("manufacturer_names", []))
+        assert meta.model_names == tuple(obj.pop("model_names", []))
+        assert meta.release_notes == obj.pop("release_notes", None)
+        assert meta.min_hardware_version == obj.pop("min_hardware_version", None)
+        assert meta.max_hardware_version == obj.pop("max_hardware_version", None)
+        assert meta.min_current_file_version == obj.pop(
+            "min_current_file_version", None
+        )
+        assert meta.max_current_file_version == obj.pop(
+            "max_current_file_version", None
+        )
+        assert meta.specificity == obj.pop("specificity", None)
+        assert meta.release_url == obj.pop("release_url", None)
```
This is expected.

TheJulianJES commented on zigpy/ota/providers.py:
```diff
@@ -660,6 +668,89 @@ async def _load_index(
+
+        # Extract the index URL from the version file
+        # Format: {"schemas": {"zigpy_v1": {"version": "...", "url": "..."}}}
+        index_url = version_data["schemas"]["zigpy_v1"]["url"]
```
I've discussed this here https://github.com/zigpy/zigpy/pull/1731#issuecomment-3694498506 and decided not to implement this, for now. If we fail here, it's only logged at `debug` level anyway, so the user will basically never see it either way.

We could adjust this in the future to have a clearer error message, but realistically, I don't think this is a real issue.

TheJulianJES committed:
Add empty line to end of index test file

TheJulianJES commented on zigpy/ota/providers.py:
```diff
@@ -660,6 +668,89 @@ async def _load_index(
+        if url in (True, None):
+            self.channel = channel or self.DEFAULT_CHANNEL
+            if self.channel not in self.SUPPORTED_CHANNELS:
+                raise ValueError(
+                    f"Invalid channel '{self.channel}'. Must be one of: {self.SUPPORTED_CHANNELS}"
```
The full error message for an invalid channel is as following. I don't think we really need to change this..
`voluptuous.error.MultipleInvalid: value must be one of ['beta', 'dev', 'stable'] for dictionary value @ data['ota']['extra_providers'][0]['channel'] `

TheJulianJES commented:
@puddly I'd merge this to get into the HA beta tomorrow, unless you say we shouldn't. The actual implementation is similar to other zigpy OTA providers for the most part. It's also tested with an actual device and works as expected.

The zigpy-ota repo is now live as well, with an announcement post: https://github.com/zigpy/zigpy-ota/discussions/2

puddly commented:
Awesome! Let's get it in. We can iterate during the beta period but I really don't see much tweaking to do, it looks really solid.
